controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -0,0 +1,291 @@
+
+    // root/scope/stream/<queueNum>/sealed
+    // root/scope/stream/<queueNum>/entities
+    public ZkOrderedStore(String ordererName, ZKStoreHelper storeHelper, Executor executor) {
+        // default roll over is after 90% or 1.8 billion entries have been added to a set. 
+        this(ordererName, storeHelper, executor, (Integer.MAX_VALUE / 10) * 9);
+    }
+
+    @VisibleForTesting
+    ZkOrderedStore(String ordererName, ZKStoreHelper storeHelper, Executor executor, int rollOverAfter) {
+        this.ordererName = ordererName;
+        this.storeHelper = storeHelper;
+        this.executor = executor;
+        this.rollOverAfter = rollOverAfter;
+    }
+
+    /**
+     * Method to add new entity to the ordered set. Note: Same entity could be added to the set multiple times.
+     * Entities are added to the latest set for the stream. If the set has exhausted positions allowed for by rollOver limit,
+     * a new successor set is initiated and the entry is added to the new set.
+     * Any entries with positions higher than rollover values under a set are ignored and eventually purged. 
+     * 
+     * @param scope scope
+     * @param stream stream 
+     * @param entity entity to add
+     * @return CompletableFuture which when completed returns the position where the entity is added to the set. 
+     */
+    public CompletableFuture<Long> addEntity(String scope, String stream, String entity) {
+        // add persistent sequential node to the latest queue number 
+        // if queueNum is sealed, increment queue number and write the entity there. 
+        return getLatestQueue(scope, stream)
+                .thenCompose(latestQueueNum ->
+                        storeHelper.createPersistentSequentialZNode(getEntitySequentialPath(scope, stream, latestQueueNum),
+                                entity.getBytes(Charsets.UTF_8))
+                                   .thenCompose(positionPath -> {
+                                       int position = getPositionFromPath(positionPath);
+                                       if (position > rollOverAfter) {
+                                           // if newly created position exceeds rollover limit, we need to delete that entry 
+                                           // and roll over. 
+                                           // 1. delete newly created path
+                                           return storeHelper.deletePath(positionPath, false)
+                                                             // 2. seal latest queue
+                                                             .thenCompose(v -> storeHelper.createZNodeIfNotExist(
+                                                                     getQueueSealedPath(scope, stream, latestQueueNum)))
+                                                             // 3. call addEntity recursively
+                                                             .thenCompose(v -> addEntity(scope, stream, entity))
```

Comment:
The flow implemented makes it a bit hard to do it, but one option to consider is to create the next last queue and add the entity atomically to avoid the recursive call. ZK provides such a feature:

https://zookeeper.apache.org/doc/r3.4.13/api/org/apache/zookeeper/Transaction.html

Reply:
Actually, we could use transactions, but the recursive call only happens in case the current "set" of almost `0.9 * Integer.Max` entries are exhausted. So the recursion will kick in few times after 1.8Billion transactions are created and committed. so i could open an issue for using zk transactions, but keep things as is for now.

Reply:
opened new issue #3570

Reply:
Hi @fpj & @shiveshr , I opened new pr #4140 to this discussion.

Some questions were discovered when implementing. Thus I'd like to ask there which operations you want to do in a transaction? Note that we provide fault tolerance if concurrent seal node/entity directory creation occurs, but this is hard(even impossible now) to implement when using ZooKeeper's Transaction.

Reply:
Besides, if our original purpose is to avoid recursive calls, I provide another possibly solution without necessary involve transaction in #3570 .